### PR TITLE
Add flags to support CA generation for Connect

### PR DESCRIPTION
**Pull request by i0rek**
There is one caveat: you have to know your clusterID in order to be able to generate a proper Connect CA. There is at least one way to get that: extract it from an existing Consul Connect CA from the very same cluster.

### CODE REVIEW

File: tlsutil/generate.go [hunk 1]
```diff
@@ -32,18 +33,61 @@ func GeneratePrivateKey() (crypto.Signer, string, error) {
...
 
+type CAOpts struct {
+	Signer      crypto.Signer
+	Serial      *big.Int
+	ClusterID   string
+	Days        int
+	Constraints []string
+	Domain      string
+	Name string
+}
```
Comment (i0rek) [inline tlsutil/generate.go:44]: New struct to make it less confusing when passing options to `GenerateCA`.

File: tlsutil/generate.go [hunk 2]
```diff
@@ -32,18 +33,61 @@ func GeneratePrivateKey() (crypto.Signer, string, error) {
...
+	days := opts.Days
+	if opts.Days == 0 {
+		days = 365
+	}
+
+	var uris []*url.URL
+	if opts.ClusterID != "" {
+		spiffeID := connect.SpiffeIDSigning{ClusterID: opts.ClusterID, Domain: opts.Domain}
+		uris = []*url.URL{spiffeID.URI()}
+	}
```
Comment (i0rek) [inline tlsutil/generate.go:85]: Bunch of defaults, that are being used instead of the null values of `CAOpts`

File: tlsutil/generate.go [hunk 3]
```diff
@@ -62,23 +106,23 @@ func GenerateCA(signer crypto.Signer, sn *big.Int, days int, constraints []strin
...
 
 	var buf bytes.Buffer
 	err = pem.Encode(&buf, &pem.Block{Type: "CERTIFICATE", Bytes: bs})
 	if err != nil {
-		return "", fmt.Errorf("error encoding private key: %s", err)
+		return "", "", fmt.Errorf("error encoding private key: %s", err)
 	}
 
-	return buf.String(), nil
+	return buf.String(), pk, nil
```
Comment (i0rek) [inline tlsutil/generate.go:125]: Also return the encoded private key, since it no longer mandatory to pass in.

File: tlsutil/generate.go [hunk 4]
```diff
@@ -62,23 +106,23 @@ func GenerateCA(signer crypto.Signer, sn *big.Int, days int, constraints []strin
 		SubjectKeyId:          id,
 	}
 
-	if len(constraints) > 0 {
+	if len(opts.Constraints) > 0 {
 		template.PermittedDNSDomainsCritical = true
-		template.PermittedDNSDomains = constraints
+		template.PermittedDNSDomains = opts.Constraints
```
Comment (mkeeler) [inline tlsutil/generate.go:111]: Should constraints here be named something like `DNSNames` or similar? Seems like that would be more descriptive of how the opt is used.

Comment (i0rek) [reply tlsutil/generate.go]: I opted for `PermittedDNSDomains` because why not...

**Review by mkeeler — APPROVED** [review]
LGTM

**Pull request merged**